src/core/net/ip6.cpp

```diff
@@ -1273,22 +1271,43 @@ otError Ip6::HandleDatagram(Message &aMessage, Netif *aNetif, const void *aLinkM
         hopLimit = header.GetHopLimit();
         aMessage.Write(Header::kHopLimitFieldOffset, hopLimit);
 
-#if OPENTHREAD_CONFIG_UNSECURE_TRAFFIC_MANAGED_BY_STACK_ENABLE
         // check whether source port is an unsecure port
+        if (aFromNcpHost && nextHeader == kProtoIcmp6)
+        {
+            Icmp::Header icmp;
+
+            SuccessOrExit(error = aMessage.Read(aMessage.GetOffset(), icmp));
+            VerifyOrExit(
+                icmp.GetType() != Icmp::Header::kTypeDstUnreach && icmp.GetType() != Icmp::Header::kTypePacketToBig &&
+                    icmp.GetType() != Icmp::Header::kTypeTimeExceeded &&
+                    icmp.GetType() != Icmp::Header::kTypeEchoRequest && icmp.GetType() != Icmp::Header::kTypeEchoReply,
```

Comment:
nit: make a constant list of ICMPv6 types and compare `icmp.GetType() ` with each element instead of hardcoding all types?